core/src/sigverifier/bls_sigverifier.rs

```diff
@@ -25,21 +39,224 @@ impl SigVerifier for BLSSigVerifier {
         packet_batches: Vec<PacketBatch>,
     ) -> Result<(), SigVerifyServiceError<Self::SendType>> {
         // TODO(wen): just a placeholder without any batching.
-        if let Some(sender) = &self.sender {
-            packet_batches.iter().for_each(|batch| {
-                batch.iter().for_each(|packet| {
-                    if let Ok(message) = packet.deserialize_slice::<BLSMessage, _>(..) {
-                        sender.send(message).unwrap();
+        let mut packet_received = 0;
+        let mut bls_messages_sent = 0;
+        let mut bls_messages_malformed = 0;
+        let mut sent_failed = 0;
+        packet_batches.iter().for_each(|batch| {
+            batch.iter().for_each(|packet| {
+                packet_received += 1;
+                match packet.deserialize_slice::<BLSMessage, _>(..) {
+                    Ok(message) => match self.sender.try_send(message) {
+                        Ok(()) => {
+                            bls_messages_sent += 1;
+                        }
+                        Err(TrySendError::Full(_)) => {
+                            warn!("BLS message channel is full, dropping message");
+                            sent_failed += 1;
+                        }
+                        Err(TrySendError::Disconnected(_)) => {
+                            // There is no hope to recover, restart the validator.
+                            panic!("BLS message channel is disconnected");
```

Comment:
how does the shutdown sequencing work? Are we careful to shut down this service before the receiver?

Reply:
sorry for the late review, we should bubble up the error to sigverify stage so that send errors are properly handled here for shutdown:
https://github.com/anza-xyz/agave/blob/6b566ba9cd58a4cb08015744f9db1236ff2f2458/core/src/sigverify_stage.rs#L415